File: C/377-combination-sum-iv/main.go
package main

import "fmt"

func combinationSum4(nums []int, target int) int {
	buf := make([]int, target+1)
	buf[0] = 1
	for i := 0; i <= target; i++ {
		if buf[i] == 0 {
			continue
		}
		for j := 0; j < len(nums); j++ {
			if i+nums[j] > target {
				continue
			}
			buf[i+nums[j]] += buf[i]
		}
	}
	return buf[target]
}

func main() {
	fmt.Println(combinationSum4([]int{1, 2, 3}, 4))
}

File: C/377-combination-sum-iv/main.rs
struct Solution {}

impl Solution {
  pub fn combination_sum4(nums: Vec<i32>, target: i32) -> i32 {
    let mut dp: Vec<i32> = vec![0; target as usize + 1];

    dp[0] = 1;

    (1..=target).for_each(|i| {
      nums.iter().for_each(|&num| {
        if i >= num {
          dp[i as usize] += dp[(i - num) as usize];
        }
      });
    });

    dp[target as usize]
  }
}
